core/trino-main/src/main/java/io/trino/cost/ComparisonStatsCalculator.java

```diff
@@ -239,4 +253,124 @@ private static PlanNodeStatsEstimate estimateExpressionNotEqualToExpression(
+        switch (operator) {
+            case LESS_THAN:
+            case LESS_THAN_OR_EQUAL:
+                return estimateExpressionLessThanOrEqualToExpression(
+                        inputStatistics,
+                        leftExpressionStatistics,
+                        leftExpressionSymbol,
+                        rightExpressionStatistics,
+                        rightExpressionSymbol,
+                        nullsFilterFactor);
+            case GREATER_THAN:
+            case GREATER_THAN_OR_EQUAL:
+                return estimateExpressionLessThanOrEqualToExpression(
+                        inputStatistics,
+                        rightExpressionStatistics,
+                        rightExpressionSymbol,
+                        leftExpressionStatistics,
+                        leftExpressionSymbol,
+                        nullsFilterFactor);
+            default:
+                throw new IllegalArgumentException("Unsupported inequality operator " + operator);
+        }
+    }
+
+    private static PlanNodeStatsEstimate estimateExpressionLessThanOrEqualToExpression(
+            PlanNodeStatsEstimate inputStatistics,
+            SymbolStatsEstimate leftExpressionStatistics,
+            Optional<Symbol> leftExpressionSymbol,
+            SymbolStatsEstimate rightExpressionStatistics,
+            Optional<Symbol> rightExpressionSymbol,
+            double nullsFilterFactor)
+    {
+        StatisticRange leftRange = StatisticRange.from(leftExpressionStatistics);
+        StatisticRange rightRange = StatisticRange.from(rightExpressionStatistics);
+        // left is always greater than right, no overlap
+        if (leftRange.getLow() > rightRange.getHigh()) {
+            return inputStatistics.mapOutputRowCount(rowCount -> 0.0);
+        }
+        // left is always lesser than right
+        if (leftRange.getHigh() < rightRange.getLow()) {
+            PlanNodeStatsEstimate.Builder estimate = PlanNodeStatsEstimate.buildFrom(inputStatistics);
+            leftExpressionSymbol.ifPresent(symbol -> estimate.addSymbolStatistics(
+                    symbol,
+                    leftExpressionStatistics.mapNullsFraction(nullsFraction -> 0.0)));
+            rightExpressionSymbol.ifPresent(symbol -> estimate.addSymbolStatistics(
+                    symbol,
+                    rightExpressionStatistics.mapNullsFraction(nullsFraction -> 0.0)));
+            return estimate.setOutputRowCount(inputStatistics.getOutputRowCount() * nullsFilterFactor)
+                    .build();
+        }
+
+        PlanNodeStatsEstimate.Builder estimate = PlanNodeStatsEstimate.buildFrom(inputStatistics);
+        double leftOverlappingRangeFraction = leftRange.overlapPercentWith(rightRange);
+        double leftAlwaysLessRangeFraction;
+        if (leftRange.getLow() < rightRange.getLow()) {
+            leftAlwaysLessRangeFraction = min(
```

Comment:
this is wrong. For
```
left: [0, 3]
right: [0, 6]
```
pairs of:
`left: [0,3] right:[3, 6]` should be part of `leftAlwaysLessRangeFraction`